packages/trader/src/Modules/Account/Sections/Security/LoginHistory/login-history.jsx

```diff
@@ -0,0 +1,169 @@
+);
+
+const parseUA = user_agent => {
+    const lookup = [
+        { name: 'Edge', regex: /(?:edge|edga|edgios|edg)\/([\d\w.-]+)/i },
+        { name: 'SeaMonkey', regex: /seamonkey\/([\d\w.-]+)/i },
+        { name: 'Opera', regex: /(?:opera|opr)\/([\d\w.-]+)/i },
+        { name: 'Chromium', regex: /(?:chromium|crios)\/([\d\w.-]+)/i },
+        { name: 'Chrome', regex: /chrome\/([\d\w.-]+)/i },
+        { name: 'Safari', regex: /version\/([\d\w.-]+)/i },
+        { name: 'IE', regex: /msie\s([\d.]+[\d])/i },
+        { name: 'IE', regex: /trident\/\d+\.\d+;.*[rv:]+(\d+\.\d)/i },
+        { name: 'Firefox', regex: /firefox\/([\d\w.-]+)/i },
+        { name: 'Binary app', regex: /binary\.com V([\d.]+)/i },
+    ];
+    for (let i = 0; i < lookup.length; i++) {
+        const info = lookup[i];
+        const match = user_agent.match(info.regex);
+        if (match !== null) {
+            return {
+                name: info.name,
+                version: match[1],
+            };
+        }
+    }
+    return null;
+};
+
+class LoginHistory extends React.Component {
+    // data keys+values
+    fields = {
+        date: localize('Date and time'),
+        action: localize('Action'),
+        browser: localize('Browser'),
+        ip: localize('IP address'),
+        status: localize('Status'),
+    };
+    // state
+    state = {
+        is_loading: true,
+        fetch_limit: 12, // TODO: put it in constants or configs
+        error: '',
+        data: [],
+    };
+    // methods
+    getLoginHistoryColumnsTemplate = () =>
+        Object.keys(this.fields).map(key => ({ title: this.fields[key], col_index: key }));
+
+    getFormattedData(login_history) {
+        const data = [];
+        for (let i = 0; i < this.state.fetch_limit; i++) {
+            data[i] = {};
+            const environment = login_history[i].environment;
+            const environment_split = environment.split(' ');
+            const date = environment_split[0];
+            const time = environment_split[1].replace('GMT', '');
+            const date_time = convertDateFormat(`${date} ${time}`, 'D-MMMM-YY hh:mm:ss', 'YYYY-MM-DD hh:mm:ss');
+            data[i].date = `${date_time} GMT`;
+            data[i].action = login_history[i].action;
+            const user_agent = environment.substring(environment.indexOf('User_AGENT'), environment.indexOf('LANG'));
+            const ua = parseUA(user_agent);
+            data[i].browser = ua ? `${ua.name} ${ua.version}` : '';
+            data[i].ip = environment_split[2].split('=')[1];
+            data[i].status = login_history[i].status === 1 ? localize('Success') : localize('Failure');
+            data[i].id = i;
+        }
+        return data;
+    }
+
+    static getLoginHistory(limit) {
+        return new Promise(resolve => {
+            WS.authorized.loginHistory(limit).then(data => {
+                if (data.error) resolve({ api_initial_load_error: data.error.message });
+                resolve(data);
```

Comment:
Do we need an else here?
```suggestion
                else resolve(data);
```

Reply:
Actually works without an `else`

Reply:
ah I see because you cannot first resolve() and then reject() a promise?